// csv-table-parser/dest/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const array_1 = require("./array");
const object_1 = require("./object");
const shape_1 = require("./shape");
const DefaultOptions = {
    separator: ',',
    keys: [],
    type: 'object',
    trim: true,
    convertNumber: true,
    convertBoolean: true,
    defaultValue: '',
    ignoreRow: {
        lackElements: true,
        emptyRow: true
    },
    startRow: 0,
    startColumn: 0,
    numberOfColumn: null
};
exports.createOptions = (options) => {
    return Object.assign({}, DefaultOptions, options, { ignoreRow: Object.assign({}, DefaultOptions.ignoreRow, options.ignoreRow) });
};
exports.parser = (tsvString, options = {}) => {
    const fixedOptions = exports.createOptions(options);
    const rows = shape_1.trimUnnecessaryElement(tsvString.split('\n'), fixedOptions.startRow);
    return fixedOptions.type === 'object'
        ? object_1.convertCsvToObject(rows, fixedOptions)
        : array_1.convertCsvToArray(rows, fixedOptions);
};
exports.default = exports.parser;
