// csv-table-parser/dest/shape.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const isNumber = require("is-number");
exports.shapeData = (str, options) => {
    const { trim } = options;
    const value = trim ? exports.trimString(str) : str;
    return exports.convertStringToCorrectType(value, options);
};
exports.isBoolean = (value) => {
    return value === 'true' || value === 'false';
};
exports.convertStringToCorrectType = (str, options) => {
    const { convertNumber, convertBoolean, defaultValue } = options;
    if (isNumber(str) && convertNumber)
        return parseInt(str);
    if (exports.isBoolean(str) && convertBoolean)
        return str === 'true';
    if (str === 'null')
        return null;
    if (str === 'undefined')
        return undefined;
    return str === '' ? defaultValue : str;
};
exports.trimString = (str) => {
    return str.replace(/^\s*/, '').replace(/\s*$/, '');
};
exports.trimUnnecessaryElement = (elements, startPosition) => {
    let i = 0;
    while (i < startPosition) {
        elements.splice(0, 1);
        i++;
    }
    return elements;
};
exports.isNeededElement = (options, index) => {
    const { startColumn, numberOfColumn } = options;
    if (index < startColumn)
        return false;
    if (!!numberOfColumn && numberOfColumn <= index)
        return false;
    return true;
};
